### ISSUE TITLE
Output plugin not able to reconnect when RMQ returns

### ISSUE THREAD

**Issue by ppf2** [logstash-plugins/logstash-output-rabbitmq#52]
Have 2 environments, 1 environment LS is able to reconnect when RMQ resumes service, another environment is not able to do so.

Reproducible steps in the problem environment:

- Ensure Logstash is stopped
- Ensure output is stopped (RMQ in this case)
- Start Logstash, wait for it to start logging about inability to connect to RMQ
- Make RMQ available again
- Logstash attempts to connect, but throws an IOException and fails to automatically reconnect.

Attempted to reproduce locally but not able to so far.  I have rabbitmq set up with a virtual host + user/password authentication (with an exchange created tied to the virtual host with permissions to the guest user) and running LS 2.2.4 as well.

```
rabbitmq {
                exchange => "test"
                exchange_type => "direct"
                host => "localhost"
                vhost => "test"
        key => "logstash-rk"
        user => "guest"
        password => "guest"
```

```
Connection to localhost:5672 refused {:exception=>"MarchHare::ConnectionRefused", :backtrace=>["/Users/User1/Downloads/logstash-2.2.4/vendor/bundle/jruby/1.9/gems/march_hare-2.15.0-java/lib/march_hare/session.rb:469:in `converting_rjc_exceptions_to_ruby'", "/Users/User1/Downloads/logstash-2.2.4/vendor/bundle/jruby/1.9/gems/march_hare-2.15.0-java/lib/march_hare/session.rb:500:in `new_connection_impl'", "/Users/User1/Downloads/logstash-2.2.4/vendor/bundle/jruby/1.9/gems/march_hare-2.15.0-java/lib/march_hare/session.rb:136:in `initialize'", "/Users/User1/Downloads/logstash-2.2.4/vendor/bundle/jruby/1.9/gems/march_hare-2.15.0-java/lib/march_hare/session.rb:109:in `connect'", "/Users/User1/Downloads/logstash-2.2.4/vendor/bundle/jruby/1.9/gems/march_hare-2.15.0-java/lib/march_hare.rb:20:in `connect'", "/Users/User1/Downloads/logstash-2.2.4/vendor/bundle/jruby/1.9/gems/logstash-mixin-rabbitmq_connection-2.3.1-java/lib/logstash/plugin_mixins/rabbitmq_connection.rb:140:in `connect'", "/Users/User1/Downloads/logstash-2.2.4/vendor/bundle/jruby/1.9/gems/logstash-mixin-rabbitmq_connection-2.3.1-java/lib/logstash/plugin_mixins/rabbitmq_connection.rb:97:in `connect!'", "/Users/User1/Downloads/logstash-2.2.4/vendor/bundle/jruby/1.9/gems/logstash-output-rabbitmq-3.0.9-java/lib/logstash/outputs/rabbitmq.rb:40:in `register'", "/Users/User1/Downloads/logstash-2.2.4/vendor/bundle/jruby/1.9/gems/logstash-core-2.2.4-java/lib/logstash/output_delegator.rb:75:in `register'", "/Users/User1/Downloads/logstash-2.2.4/vendor/bundle/jruby/1.9/gems/logstash-core-2.2.4-java/lib/logstash/pipeline.rb:173:in `start_workers'", "org/jruby/RubyArray.java:1613:in `each'", "/Users/User1/Downloads/logstash-2.2.4/vendor/bundle/jruby/1.9/gems/logstash-core-2.2.4-java/lib/logstash/pipeline.rb:173:in `start_workers'", "/Users/User1/Downloads/logstash-2.2.4/vendor/bundle/jruby/1.9/gems/logstash-core-2.2.4-java/lib/logstash/pipeline.rb:126:in `run'", "/Users/User1/Downloads/logstash-2.2.4/vendor/bundle/jruby/1.9/gems/logstash-core-2.2.4-java/lib/logstash/agent.rb:210:in `execute'", "/Users/User1/Downloads/logstash-2.2.4/vendor/bundle/jruby/1.9/gems/logstash-core-2.2.4-java/lib/logstash/runner.rb:90:in `run'", "org/jruby/RubyProc.java:281:in `call'", "/Users/User1/Downloads/logstash-2.2.4/vendor/bundle/jruby/1.9/gems/logstash-core-2.2.4-java/lib/logstash/runner.rb:95:in `run'", "org/jruby/RubyProc.java:281:in `call'", "/Users/User1/Downloads/logstash-2.2.4/vendor/bundle/jruby/1.9/gems/stud-0.0.22/lib/stud/task.rb:24:in `initialize'"], :level=>:error, :file=>"logstash/plugin_mixins/rabbitmq_connection.rb", :line=>"99", :method=>"connect!"}
Connecting to RabbitMQ. Settings: {:vhost=>"test", :host=>"localhost", :port=>5672, :user=>"guest", :automatic_recovery=>true, :pass=>"guest", :timeout=>0, :heartbeat=>0} {:level=>:debug, :file=>"logstash/plugin_mixins/rabbitmq_connection.rb", :line=>"138", :method=>"connect"}
Connected to RabbitMQ at localhost {:level=>:info, :file=>"logstash/plugin_mixins/rabbitmq_connection.rb", :line=>"145", :method=>"connect"}
Declaring an exchange {:name=>"test", :type=>"direct", :durable=>true, :level=>:debug, :file=>"logstash/plugin_mixins/rabbitmq_connection.rb", :line=>"124", :method=>"declare_exchange!"}
Exchange declared {:level=>:debug, :file=>"logstash/plugin_mixins/rabbitmq_connection.rb", :line=>"127", :method=>"declare_exchange!"}
Will start workers for output {:worker_count=>4, :class=>LogStash::Outputs::RabbitMQ, :level=>:debug, :file=>"logstash/output_delegator.rb", :line=>"77", :method=>"register"}
```

This reconnects fine, but for the other environment, an IOException is thrown when it tries to reconnect. 

```
{: timestamp => "2016-06-30T13:04:01.992000-0400", : message => "Connection to <redacted>:2997 refused", : exception => "MarchHare::ConnectionRefused", : backtrace => ["/opt/logstash/vendor/bundle/jruby/1.9/gems/march_hare-2.15.0-java/lib/march_hare/session.rb:469:in `converting_rjc_exceptions_to_ruby'", "/opt/logstash/vendor/bundle/jruby/1.9/gems/march_hare-2.15.0-java/lib/march_hare/session.rb:500:in `new_connection_impl'", "/opt/logstash/vendor/bundle/jruby/1.9/gems/march_hare-2.15.0-java/lib/march_hare/session.rb:136:in `initialize'", "/opt/logstash/vendor/bundle/jruby/1.9/gems/march_hare-2.15.0-java/lib/march_hare/session.rb:109:in `connect'", "/opt/logstash/vendor/bundle/jruby/1.9/gems/march_hare-2.15.0-java/lib/march_hare.rb:20:in `connect'", "/opt/logstash/vendor/bundle/jruby/1.9/gems/logstash-mixin-rabbitmq_connection-2.3.1-java/lib/logstash/plugin_mixins/rabbitmq_connection.rb:140:in `connect'", "/opt/logstash/vendor/bundle/jruby/1.9/gems/logstash-mixin-rabbitmq_connection-2.3.1-java/lib/logstash/plugin_mixins/rabbitmq_connection.rb:97:in `connect!'", "/opt/logstash/vendor/bundle/jruby/1.9/gems/logstash-output-rabbitmq-3.0.9-java/lib/logstash/outputs/rabbitmq.rb:40:in `register'", "/opt/logstash/vendor/bundle/jruby/1.9/gems/logstash-core-2.2.4-java/lib/logstash/output_delegator.rb:75:in `register'", "/opt/logstash/vendor/bundle/jruby/1.9/gems/logstash-core-2.2.4-java/lib/logstash/pipeline.rb:173:in `start_workers'", "org/jruby/RubyArray.java:1613:in `each'", "/opt/logstash/vendor/bundle/jruby/1.9/gems/logstash-core-2.2.4-java/lib/logstash/pipeline.rb:173:in `start_workers'", "/opt/logstash/vendor/bundle/jruby/1.9/gems/logstash-core-2.2.4-java/lib/logstash/pipeline.rb:126:in `run'", "/opt/logstash/vendor/bundle/jruby/1.9/gems/logstash-core-2.2.4-java/lib/logstash/agent.rb:210:in `execute'", "/opt/logstash/vendor/bundle/jruby/1.9/gems/logstash-core-2.2.4-java/lib/logstash/runner.rb:90:in `run'", "org/jruby/RubyProc.java:281:in `call'", "/opt/logstash/vendor/bundle/jruby/1.9/gems/logstash-core-2.2.4-java/lib/logstash/runner.rb:95:in `run'", "org/jruby/RubyProc.java:281:in `call'", "/opt/logstash/vendor/bundle/jruby/1.9/gems/stud-0.0.22/lib/stud/task.rb:24:in `initialize'"], : level => : error, : file => "logstash/plugin_mixins/rabbitmq_connection.rb", : line => "99", : method => "connect!"
} {: timestamp => "2016-06-30T13:04:02.994000-0400", : message => "Connecting to RabbitMQ. Settings: {:vhost=>\"logstash-vh\", :host=>\"<redacted>\", :port=>2997, :user=>\"logstash\", :automatic_recovery=>true, :pass=>\"logstash\", :timeout=>0, :heartbeat=>0}", : level => : debug, : file => "logstash/plugin_mixins/rabbitmq_connection.rb", : line => "138", : method => "connect"
} {: timestamp => "2016-06-30T13:04:04.015000-0400", : message => "An unexpected error occurred!", : error => java.io.IOException, : class => "Java::JavaIo::IOException", : backtrace => ["com.rabbitmq.client.impl.AMQChannel.wrap(com/rabbitmq/client/impl/AMQChannel.java:106)", "com.rabbitmq.client.impl.AMQChannel.wrap(com/rabbitmq/client/impl/AMQChannel.java:102)", "com.rabbitmq.client.impl.AMQConnection.start(com/rabbitmq/client/impl/AMQConnection.java:350)", "com.rabbitmq.client.ConnectionFactory.newConnection(com/rabbitmq/client/ConnectionFactory.java:648)", "com.rabbitmq.client.ConnectionFactory.newConnection(com/rabbitmq/client/ConnectionFactory.java:678)", "java.lang.reflect.Method.invoke(java/lang/reflect/Method.java:498)", "RUBY.new_connection_impl(/opt/logstash/vendor/bundle/jruby/1.9/gems/march_hare-2.15.0-java/lib/march_hare/session.rb:505)", "org.jruby.RubyProc.call(org/jruby/RubyProc.java:281)", "RUBY.converting_rjc_exceptions_to_ruby(/opt/logstash/vendor/bundle/jruby/1.9/gems/march_hare-2.15.0-java/lib/march_hare/session.rb:467)", "RUBY.new_connection_impl(/opt/logstash/vendor/bundle/jruby/1.9/gems/march_hare-2.15.0-java/lib/march_hare/session.rb:500)", "RUBY.initialize(/opt/logstash/vendor/bundle/jruby/1.9/gems/march_hare-2.15.0-java/lib/march_hare/session.rb:136)", "RUBY.connect(/opt/logstash/vendor/bundle/jruby/1.9/gems/march_hare-2.15.0-java/lib/march_hare/session.rb:109)", "RUBY.connect(/opt/logstash/vendor/bundle/jruby/1.9/gems/march_hare-2.15.0-java/lib/march_hare.rb:20)", "RUBY.connect(/opt/logstash/vendor/bundle/jruby/1.9/gems/logstash-mixin-rabbitmq_connection-2.3.1-java/lib/logstash/plugin_mixins/rabbitmq_connection.rb:140)", "RUBY.connect!(/opt/logstash/vendor/bundle/jruby/1.9/gems/logstash-mixin-rabbitmq_connection-2.3.1-java/lib/logstash/plugin_mixins/rabbitmq_connection.rb:97)", "RUBY.register(/opt/logstash/vendor/bundle/jruby/1.9/gems/logstash-output-rabbitmq-3.0.9-java/lib/logstash/outputs/rabbitmq.rb:40)", "RUBY.register(/opt/logstash/vendor/bundle/jruby/1.9/gems/logstash-core-2.2.4-java/lib/logstash/output_delegator.rb:75)", "RUBY.start_workers(/opt/logstash/vendor/bundle/jruby/1.9/gems/logstash-core-2.2.4-java/lib/logstash/pipeline.rb:173)", "org.jruby.RubyArray.each(org/jruby/RubyArray.java:1613)", "RUBY.start_workers(/opt/logstash/vendor/bundle/jruby/1.9/gems/logstash-core-2.2.4-java/lib/logstash/pipeline.rb:173)", "RUBY.run(/opt/logstash/vendor/bundle/jruby/1.9/gems/logstash-core-2.2.4-java/lib/logstash/pipeline.rb:126)", "RUBY.execute(/opt/logstash/vendor/bundle/jruby/1.9/gems/logstash-core-2.2.4-java/lib/logstash/agent.rb:210)", "RUBY.run(/opt/logstash/vendor/bundle/jruby/1.9/gems/logstash-core-2.2.4-java/lib/logstash/runner.rb:90)", "org.jruby.RubyProc.call(org/jruby/RubyProc.java:281)", "RUBY.run(/opt/logstash/vendor/bundle/jruby/1.9/gems/logstash-core-2.2.4-java/lib/logstash/runner.rb:95)", "org.jruby.RubyProc.call(org/jruby/RubyProc.java:281)", "RUBY.initialize(/opt/logstash/vendor/bundle/jruby/1.9/gems/stud-0.0.22/lib/stud/task.rb:24)", "java.lang.Thread.run(java/lang/Thread.java:745)"], : level => : warn, : file => "logstash/agent.rb", : line => "221", : method => "execute"
}
```

```
output  {
        rabbitmq {
                exchange => "logstash-ex"
                exchange_type => "direct"
                host => "rabbitmq_host"
                vhost => "logstash-vh"
        key => "logstash-rk"
        port => "2997"
        user => "logstash"
        password => "password"
                }
        }
```

**Comment by andrewvc** [comment]
@michaelklishin are vhosts currently broken in `march_hare` and rabbitmq java? 

Trying the following with v3.6.1 of the server on OSX I get:

```
require 'march_hare'
irb(main):001:0> require 'march_hare'
=> true
irb(main):002:0> MarchHare.connect(:exchange => "logstash-ex", :exchange_type => "direct", :vhost => "logstash-vh")
Java::JavaIo::IOException:
	from com.rabbitmq.client.impl.AMQChannel.wrap(com/rabbitmq/client/impl/AMQChannel.java:106)
	from com.rabbitmq.client.impl.AMQChannel.wrap(com/rabbitmq/client/impl/AMQChannel.java:102)
	from com.rabbitmq.client.impl.AMQChannel.exnWrappingRpc(com/rabbitmq/client/impl/AMQChannel.java:124)
	from com.rabbitmq.client.impl.AMQConnection.start(com/rabbitmq/client/impl/AMQConnection.java:378)
	from com.rabbitmq.client.ConnectionFactory.newConnection(com/rabbitmq/client/ConnectionFactory.java:647)
	from com.rabbitmq.client.ConnectionFactory.newConnection(com/rabbitmq/client/ConnectionFactory.java:677)
	from java.lang.reflect.Method.invoke(java/lang/reflect/Method.java:497)
	from RUBY.block in new_connection_impl(/Users/andrewvc/.rbenv/versions/jruby-9.0.3.0/lib/ruby/gems/shared/gems/march_hare-2.16.0-java/lib/march_hare/session.rb:505)
	from org.jruby.RubyProc.call(org/jruby/RubyProc.java:318)
	from RUBY.converting_rjc_exceptions_to_ruby(/Users/andrewvc/.rbenv/versions/jruby-9.0.3.0/lib/ruby/gems/shared/gems/march_hare-2.16.0-java/lib/march_hare/session.rb:467)
	from RUBY.new_connection_impl(/Users/andrewvc/.rbenv/versions/jruby-9.0.3.0/lib/ruby/gems/shared/gems/march_hare-2.16.0-java/lib/march_hare/session.rb:500)
	from RUBY.initialize(/Users/andrewvc/.rbenv/versions/jruby-9.0.3.0/lib/ruby/gems/shared/gems/march_hare-2.16.0-java/lib/march_hare/session.rb:136)
	from RUBY.connect(/Users/andrewvc/.rbenv/versions/jruby-9.0.3.0/lib/ruby/gems/shared/gems/march_hare-2.16.0-java/lib/march_hare/session.rb:109)
	from RUBY.connect(/Users/andrewvc/.rbenv/versions/jruby-9.0.3.0/lib/ruby/gems/shared/gems/march_hare-2.16.0-java/lib/march_hare.rb:20)
	from RUBY.<eval>((irb):2)
	from org.jruby.RubyKernel.eval(org/jruby/RubyKernel.java:978)
	from org.jruby.RubyKernel.loop(org/jruby/RubyKernel.java:1291)
	from org.jruby.RubyKernel.catch(org/jruby/RubyKernel.java:1098)
	from org.jruby.RubyKernel.catch(org/jruby/RubyKernel.java:1098)
	from RUBY.<top>(/Users/andrewvc/.rbenv/versions/jruby-9.0.3.0/bin/irb:13)irb(main):003:0>
```

thanks for the report @ppf2

**Comment by andrewvc** [comment]
OK the issue here is that we ran into https://github.com/rabbitmq/rabbitmq-java-client/issues/100 which apparently will not be fixed due to a deficiency of the AMQP protocol.

It seems that the user in question does not have perms for the vhost in question.

**Comment by andrewvc** [comment]
Since this is marked wontfix on the java client there isn't anything we can do here, closing the ticket.

**Comment by jordansissel** [comment]
I think I have an idea how to fix this. PR coming shortly.

**Comment by jordansissel** [comment]
https://github.com/logstash-plugins/logstash-mixin-rabbitmq_connection/pull/26 filed for this.

I am not sure how to reproduce this problem, but I believe my PR should fix it by causing `connect!` to retry when IOException is thrown.

**Comment by andrewvc** [comment]
Fixed with @jordansissel 's fix + the most 4.20 release of the connection mixin
